tests/zero-downtime-upgrades/README.md

```diff
@@ -0,0 +1,85 @@
+# Zero-Downtime Upgrades
+
+This document describes a test plan for testing zero-downtime upgrades of NGF.
+
+*Zero-downtime upgrades* means that during an NGF upgrade clients don't experience any
+interruptions to the traffic they send to applications exposed via NGF.
+
+## Goals
+
+- Ensure that upgrading NFG doesn't lead to any loss of traffic flowing through the data plane.
+- Ensure that after an upgrade, NGF can process changes to resources.
+
+## Non-Goals
+
+During an upgrade, Kubernetes will shut down existing NGF pods by sending a SIGTERM. If the pod doesn't terminate in 30
+seconds (the default period) , Kubernetes will send a SIGKILL.
+
+When proxying Websocket or any long-lived connections, NGINX will not terminate until
+that connection is closed by either the client or the backend. This means that unless all those connections are closed
+by clients/backends before or during an upgrade (which is highly unlikely), NGINX will not terminate, which means
+Kubernetes will kill NGINX. As a result, the clients will see the connections abruptly closed and thus experience
+downtime.
+
+As a result, we *will not* use any long-live connections in this test, because NGF cannot support zero-downtime upgrades
+in this case.
+
+## Test Environment
+
+- A Kubernetes cluster with 3 nodes on GKE
+    - Node: e2-medium (2 vCPU, 4GB memory)
+    - Enabled GKE logging.
+- Tester VMs:
```

Comment:
Is a VM a requirement for this test?

Reply:
The requirement is to generate a steady load of traffic.
VM is convenient, because it is located in the same zone, which reduced the risk of networking problems during the test. Additionally, I'm using an internal regional load balancer, so that data plane is not exposed to the outside

Reply:
Could we just send traffic from another Pod? Or do we want to ensure that we are acting as a client outside of the cluster?

Reply:
Our typically user will use a service type load balancer to expose NGF. 
In the context of this test, downtime means external clients can't access NGF.
Because of that, it is important to send traffic through a load balancer type service.